lib/models/user.js

```diff
@@ -1,10 +1,95 @@
 'use strict';
 
-const errors = require('storj-service-error-types');
 const activator = require('hat').rack(256);
 const crypto = require('crypto');
 const mongoose = require('mongoose');
 const SchemaOptions = require('../options');
+const PAYMENT_PROCESSORS = require('../constants').PAYMENT_PROCESSORS;
+const paymentProcessorAdapters = require('../graphql/payment-processor-adapters/payment-processor-adapters');
+const errors = require('storj-service-error-types');
+
+// NB: due to mongoose weirdness, `PaymentProcessor`s virtuals must be defined before
+// it is assigned as a child schema
+const PaymentProcessor = new mongoose.Schema({
+  name: {
+    type: String,
+    enum: Object.keys(PAYMENT_PROCESSORS).map((key) => (PAYMENT_PROCESSORS[key])),
+    unique: true
+  },
+  rawData: [{
+    type: mongoose.Schema.Types.Mixed
+  }],
+  // TODO: add `pre` such that only one payment processor can be default
+  default: {
+    type: Boolean
+  },
+  created: {
+    type: Date,
+    default: Date.now
+  }
+});
+
+PaymentProcessor.set('toObject', {
+  virtuals: true,
+  transform: (doc, ret) => {
+    delete ret.__v;
+    delete ret._id;
+    delete ret.rawData;
+    // delete ret.data;
+    // delete ret.adapter;
+  }
+});
+
+PaymentProcessor.virtual('adapter').get(function() {
+  return paymentProcessorAdapters[this.name];
+});
+
+PaymentProcessor.virtual('data')
+    .get(function() {
+      return this.adapter.parseData(this.rawData);
+    })
+    .set(function(data) {
+      return this.adapter.serializeData(data);
+    })
+;
+
+// TODO: maybe `defaultPaymentMethod` instead? (structure would change then...)
+PaymentProcessor.virtual('defaultCard').get(function() {
+  return this.adapter.defaultPaymentMethod(this);
+});
+
+PaymentProcessor.virtual('billingDate').get(function() {
+  return this.adapter.billingDate(this);
+});
+
+PaymentProcessor.methods.update = function(data) {
+  return this.adapter.validate()
```

Comment:
I agree, it could result in unexpected behavior down the road. I'm not sure how @bryanchriswhite would feel about rewriting for callback syntax, though. I know that most of our billing code that relies on these models is promise based and would require a refactor to work with callback pattern. I'm open-minded on the issue myself, however. If we think that callback pattern should be maintained, we can open an issue for it and address it after all of this is deployed.